### PR TITLE
HoundConfig merges default and user configurations

### PR DESCRIPTION
**Pull request by jcmorrow**
Previously, `hound_config#enabled_for?` would return `true` for any value that was not:

 - Explicitly set to false in the supplied config file
 - A language marked as beta

Additionally, HoundConfig didn't account for aliases. This could result in someone disabling an alias (e.g. `javascript`) but still running the actual linter (e.g. `jshint`). Instead of trying to perform boolean logic about what is and is not activated, this strategy just starts with a default configuration and merges the user-specified configuration into it using a series of transforms.

Thanks to @jackvnimble for the idea!

### CODE REVIEW

File: app/services/config_alias_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  ALIASES = {
+    jshint: [:javascript, :java_script],
+    coffeescript: [:coffee_script],
+  }
+
+  def initialize(config)
+    @config = config
+  end
+
+  def run
```
Comment (jcmorrow) [inline app/services/config_alias_resolver.rb]: I agree hound! 🐶

File: app/services/config_alias_resolver.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+class ConfigAliasResolver
+  ALIASES = {
+    "javascript" => "jshint",
+    "java_script" => "jshint",
+    "coffee_script" => "coffeescript"
```
Comment (gylaz) [inline app/services/config_alias_resolver.rb]: Listen to Da Hound!

File: app/models/hound_config.rb [hunk 3]
```diff
@@ -1,31 +1,30 @@
...
+    "go" => { "enabled" => true },
+    "haml" => { "enabled" => true },
+    "jscs" => { "enabled" => false },
+    "jshint" => { "enabled" => true },
+    "remark" => { "enabled" => false },
+    "python" => { "enabled" => false },
+    "ruby" => { "enabled" => true },
+    "scss" => { "enabled" => true },
+    "swift" => { "enabled" => true },
+  }.freeze
```
Comment (gylaz) [inline app/models/hound_config.rb]: Here's an idea: Since the linter names can change (yet again). How about generating a default config from the list of `LINTERS` and setting enabled to `true` (or `false` if in `BETA`)?

File: app/models/hound_config.rb [hunk 4]
```diff
@@ -1,31 +1,30 @@
...
 
   attr_reader_initialize :commit
 
   def content
-    @content ||= parse(commit.file_content(CONFIG_FILE))
+    merged_config
```
Comment (gylaz) [inline app/models/hound_config.rb]: Any value in having this method call another rather than just inlining the logic?

File: app/models/hound_config.rb [hunk 5]
```diff
@@ -34,32 +33,38 @@ def fail_on_violations?
 
   private
 
-  def parse(file_content)
-    Config::Parser.yaml(file_content) || {}
+  def resolved_aliases_config
+    @resolved_aliases_config ||= ConfigAliasResolver.new(normalized_config).run
```
Comment (gylaz) [inline app/models/hound_config.rb]: Probably don't need to memoize this, since we memoize right after this is called.

File: app/models/hound_config.rb [hunk 6]
```diff
@@ -34,32 +33,38 @@ def fail_on_violations?
...
   end
 
-  def configured?(name)
-    key = normalize_key(name)
-    config = options_for(key)
+  def normalized_config
+    @normalized_config ||= ConfigNormalizer.new(parsed_config).run
```
Comment (gylaz) [inline app/models/hound_config.rb]: Same here.

File: app/models/hound_config.rb (hunk 5 above)
Comment (jcmorrow) [inline app/models/hound_config.rb]: Good call, I was moving things around and definitely got carried away with the memoizing.

File: app/models/hound_config.rb (hunk 4 above)
Comment (jcmorrow) [inline app/models/hound_config.rb]: Nope, it used to do more and now it doesn't, inlining

File: app/models/hound_config.rb [hunk 7]
```diff
@@ -34,32 +33,38 @@ def fail_on_violations?
...
-  def options_for(name)
-    config = content[name] || {}
-    default_options_for(name).merge(config)
+  def merge(base_hash, derived_hash)
+    result = base_hash.merge(derived_hash)
+    keys_appearing_in_both = base_hash.keys & derived_hash.keys
+    keys_appearing_in_both.each do |key|
+      next unless base_hash[key].is_a?(Hash)
+      result[key] = merge(base_hash[key], derived_hash[key])
+    end
```
Comment (gylaz) [inline app/models/hound_config.rb]: I'm having a hard time parsing this (and why we need this). How about extracting this into an intention revealing method?

File: app/services/config_alias_resolver.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+class ConfigAliasResolver
```
Comment (gylaz) [inline app/services/config_alias_resolver.rb]: How about naming this service after what it does `ResolveConfigAliases`}

File: app/services/config_alias_resolver.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  }.freeze
+
+  def initialize(config)
+    @config = config
+  end
+
+  def run
+    config_with_aliases_resolved = {}
+
+    @config.each_pair do |config_key, config_value|
```
Comment (gylaz) [inline app/services/config_alias_resolver.rb]: Using `each` is more idiomatic: `@config.each do |key, value|`

Comment (gylaz) [inline app/services/config_alias_resolver.rb]: Even better would be to use `reduce` or `each_with_object`.

File: app/services/config_normalizer.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
+class ConfigNormalizer
```
Comment (gylaz) [inline app/services/config_normalizer.rb]: `NormalizeConfig`?

Comment (jcmorrow) [inline app/services/config_normalizer.rb]: Oh yeah, I guess we name them like tasks rather than nouns don't we? Sounds good to me.

File: app/services/config_alias_resolver.rb (hunk 9 above)
Comment (jcmorrow) [inline app/services/config_alias_resolver.rb]: Yeah I just switched something else from `each` to `inject` and had the same thought, these things are basically supposed to be functional so those techniques make more sense here.

File: app/models/hound_config.rb (hunk 7 above)
Comment (jcmorrow) [inline app/models/hound_config.rb]: yeah, it should be called like `deep_merge` or something like that. I'm not sure if it can get much more broken up due to the recursive nature, but I can take a shot.

File: app/services/resolve_config_aliases.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    "java_script" => "jshint",
+    "coffee_script" => "coffeescript",
+  }.freeze
+
+  def initialize(config)
+    @config = config
+  end
+
+  def run
+    @config.reduce({}) do |config_without_aliases, (config_key, config_value)|
```
Comment (gylaz) [inline app/services/resolve_config_aliases.rb]: This is actually wrong, according to thoughtbot styleguide. I didn't realize this but since we moved this under `houndci` org, we no longer get the thoughtbot rules.

File: app/models/hound_config.rb [hunk 12]
```diff
@@ -34,32 +31,34 @@ def fail_on_violations?
...
   end
 
-  def options_for(name)
-    config = content[name] || {}
-    default_options_for(name).merge(config)
+  def merge(base_hash, derived_hash)
+    result = base_hash.merge(derived_hash)
+    keys_appearing_in_both = base_hash.keys & derived_hash.keys
+    keys_appearing_in_both.each do |key|
+      next unless base_hash[key].is_a?(Hash)
```
Comment (teoljungberg) [inline app/models/hound_config.rb]: `next` inside of an `each` loop is not a ruby convention - try using a `select` or `reject` first to limit the data set.

**Comment by teoljungberg** [comment]
@jcmorrow This PR seems to use work that was done in #1143 - which makes the diff quite confusing to read.

File: app/models/hound_config.rb (hunk 12 above)
Comment (teoljungberg) [inline app/models/hound_config.rb]: This could also be refactored into a `reduce`.

File: app/models/hound_config.rb [hunk 13]
```diff
@@ -34,32 +31,34 @@ def fail_on_violations?
 
   private
 
-  def parse(file_content)
-    Config::Parser.yaml(file_content) || {}
+  def resolved_aliases_config
+    ResolveConfigAliases.new(normalized_config).run
```
Comment (teoljungberg) [inline app/models/hound_config.rb]: What do you think about using a top level `.run` method? Which is a convention in this project.

File: app/models/hound_config.rb [hunk 14]
```diff
@@ -34,32 +31,34 @@ def fail_on_violations?
...
   end
 
-  def configured?(name)
-    key = normalize_key(name)
-    config = options_for(key)
-
-    enabled?(config)
+  def normalized_config
+    NormalizeConfig.new(parsed_config).run
```
Comment (teoljungberg) [inline app/models/hound_config.rb]: Same as above.

File: app/services/normalize_config.rb [hunk 15]
```diff
@@ -0,0 +1,23 @@
+class NormalizeConfig
+  def initialize(config)
+    @config = config
+  end
+
+  def run
+    @config.inject({}) do |normalized_config, (key, value)|
```
Comment (teoljungberg) [inline app/services/normalize_config.rb]: We prefer `reduce` over `inject` - https://github.com/thoughtbot/guides/blob/5deeaaa2006a7f895f697cc99668eadaa4abe693/style/ruby/README.md.

Also - I'm unsure why this isn't being warned by Hound.

File: spec/models/hound_config_spec.rb [hunk 16]
```diff
@@ -1,138 +1,125 @@
...
+          commit = stubbed_commit(".hound.yml" => "")
+          hound_config = HoundConfig.new(commit)
+
+          expect(hound_config).not_to be_enabled_for("remark")
         end
       end
     end
 
-    context "when the given language is disabled" do
+    context "given a language that is disabled in the config file" do
```
Comment (teoljungberg) [inline spec/models/hound_config_spec.rb]: I don't think this wording needed to change.

Comment (teoljungberg) [inline spec/models/hound_config_spec.rb]: Nor the test bellow.

File: spec/models/hound_config_spec.rb [hunk 17]
```diff
@@ -173,4 +160,8 @@
       end
     end
   end
+
+  def be_enabled_for(linter_name)
```
Comment (teoljungberg) [inline spec/models/hound_config_spec.rb]: Nice refactoring!

File: spec/services/config_normalizer_spec.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+require "app/services/normalize_config"
+
+describe NormalizeConfig do
+  context "given a hash with keys containing capital letters" do
```
Comment (teoljungberg) [inline spec/services/config_normalizer_spec.rb]: Should we have a `describe "#run"` here?

File: spec/services/config_alias_resolver_spec.rb [hunk 19]
```diff
@@ -0,0 +1,16 @@
+require "app/services/resolve_config_aliases"
+
+describe ResolveConfigAliases do
+  context "when the content contains configured aliases" do
```
Comment (teoljungberg) [inline spec/services/config_alias_resolver_spec.rb]: Should we have a `describe "#run"` here?

Comment (teoljungberg) [inline spec/services/config_alias_resolver_spec.rb]: I would replace `content` with `config` in this context, especially since the class name has `config` in it.

Comment (teoljungberg) [inline spec/services/config_alias_resolver_spec.rb]: Also, the spec name doesn't match the class name.

File: spec/services/config_normalizer_spec.rb (hunk 18 above)
Comment (teoljungberg) [inline spec/services/config_normalizer_spec.rb]: The spec name does not match the class name.

File: app/services/normalize_config.rb (hunk 15 above)
Comment (jcmorrow) [inline app/services/normalize_config.rb]: Hound warned me the other way, hence the change 😕

File: spec/services/config_alias_resolver_spec.rb (hunk 19 above)
Comment (jcmorrow) [inline spec/services/config_alias_resolver_spec.rb]: Good catches all!

File: spec/models/hound_config_spec.rb (hunk 16 above)
Comment (jcmorrow) [inline spec/models/hound_config_spec.rb]: I kind of prefer the contexts to be consistent in their wording, but it's not a necessary change.

**Comment by jcmorrow** [comment]
@teoljungberg I probably should have been more clear, this is an alternative to https://github.com/houndci/hound/pull/1143. I wanted to be able to see how both solutions worked out without smashing older work with the newer approach.

File: app/models/hound_config.rb (hunk 7 above)
Comment (jcmorrow) [inline app/models/hound_config.rb]: TIL, Rails has a `deep_merge` method. http://api.rubyonrails.org/v2.3/classes/ActiveSupport/CoreExtensions/Hash/DeepMerge.html

File: app/services/normalize_config.rb [hunk 20]
```diff
@@ -0,0 +1,19 @@
+class NormalizeConfig
+  def self.run(config)
+    config.reduce({}) do |normalized_config, (key, value)|
```
Comment (jcmorrow) [inline app/services/normalize_config.rb]: @teoljungberg ❓

Comment (teoljungberg) [inline app/services/normalize_config.rb]: Beats me...

Comment (gylaz) [inline app/services/normalize_config.rb]: I've made this comment else where in this PR:
> This is actually wrong, according to thoughtbot styleguide. I didn't realize this but since we moved this under houndci org, we no longer get the thoughtbot rules.

So, we should add this rule back in to our `.ruby-style.yml`.

File: app/models/hound_config.rb [hunk 21]
```diff
@@ -1,31 +1,28 @@
...
-    coffeescript
-    go
-    haml
-    javascript
-    python
-    ruby
-    scss
-    swift
-  )
+  DEFAULT_CONFIG = Linter::Collection::LINTER_NAMES.inject({}) do |config, name|
```
Comment (gylaz) [inline app/models/hound_config.rb]: If you use `each_with_object` you can avoid needing `config` on line 12.

Comment (gylaz) [inline app/models/hound_config.rb]: Constant feels like a wrong tool for this, since we are doing some logic. Thoughts on using a class method for this instead?

File: app/services/normalize_config.rb (hunk 20 above)
Comment (gylaz) [inline app/services/normalize_config.rb]:
```yaml
Style/CollectionMethods:
  Enabled: true
  PreferredMethods:
    find: detect
    inject: reduce
    collect: map
    find_all: select
```

File: app/models/hound_config.rb (hunk 21 above)
Comment (jcmorrow) [inline app/models/hound_config.rb]: That's a good call.

File: app/models/hound_config.rb [hunk 22]
```diff
@@ -1,31 +1,27 @@
...
-    coffeescript
-    go
-    haml
-    javascript
-    python
-    ruby
-    scss
-    swift
-  )
+  DEFAULT_CONFIG =
```
Comment (teoljungberg) [inline app/models/hound_config.rb]: This also seems like it should be a method.

File: app/services/normalize_config.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
+class NormalizeConfig
+  def self.run(config)
+    new(config).normalize
```
Comment (teoljungberg) [inline app/services/normalize_config.rb]: We usually name the instance method the same as its static facade - so `.run` -> `#run`.

File: app/services/resolve_config_aliases.rb [hunk 24]
```diff
@@ -0,0 +1,26 @@
+class ResolveConfigAliases
```
Comment (teoljungberg) [inline app/services/resolve_config_aliases.rb]: I wonder about the use case of this service object - we already have a way of resolving aliases for a config using `Config::Base#linter_names`.

Couldn't we use that approach and have fewer moving parts?

File: spec/models/hound_config_spec.rb [hunk 25]
```diff
@@ -1,37 +1,44 @@
 require "spec_helper"
+require "app/models/config/parser"
+require "app/services/normalize_config"
+require "app/services/resolve_config_aliases"
 require "app/models/hound_config"
 
 describe HoundConfig do
   describe "#content" do
-    it "returns the content of the .hound.yml file" do
+    it "returns the contents of the .hound.yml file merged with the defaults" do
```
Comment (teoljungberg) [inline spec/models/hound_config_spec.rb]: I would add a comma after `file`

**Comment by teoljungberg** [comment]
I'm somewhat confused over the approach in this PR, as in we've built new functionality to replace something that we already have. So some questions off the bat:

- Doesn't the existing configuration setup work with aliases? If not - we should fix that instead of introducing new concepts. See `Config::Base#linter_names`.
- There's a bit too much going on in this PR, and it seems like they are isolated enough that it can be extracted into multiple PR's.

**Comment by jcmorrow** [comment]
@teoljungberg Our current setup doesn't really work. It **sort of** works, but it misses a lot of corner cases. The biggest thing it misses is:

Currently if I disable an alias of a default linter, it doesn't disable the linter. So if your `hound.yml` is:

    coffee_script: 
      enabled: false

coffeescript will still be run. The only reason this works:

    javascript: 
      enabled: false

is because `jshint` is under `BETA_LANGUAGES`, while `javascript` is under `LANGUAGES`, which doesn't make any sense since they are aliases for the same linter. So doing this:


    jshint:
      enabled: false

Actually wouldn't disable the jshint linter.

----

I agree that it would be nice to just make the current approach work, but it involves making the changes I made in https://github.com/houndci/hound/pull/1143 and those were much more fragile and harder to grok than the approach taken here. I'm open to breaking this up into smaller PR's. I'll see if I can start that today.

File: app/services/resolve_config_aliases.rb (hunk 24 above)
Comment (jcmorrow) [inline app/services/resolve_config_aliases.rb]: The problem with this is the problem of disabled aliases, as I mentioned in the main discussion thread. Might be worth talking over on a hangout?

File: spec/services/resolve_config_aliases_spec.rb [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+describe ResolveConfigAliases do
+  describe "#run" do
+    context "when the config contains aliases" do
+      it "renames them to the appropriate linter" do
+        config = {
+          "javascript" => { "enabled" => false },
+          "ruby" => { "enabled" => false },
+        }
+
+        expect(ResolveConfigAliases.run(config).keys).to eq(["jshint", "ruby"])
```
Comment (thorncp) [inline spec/services/resolve_config_aliases_spec.rb]: Might be good to use `match_array` here so we don't rely on the ordering of the keys in the hash.

File: spec/services/normalize_config_spec.rb [hunk 27]
```diff
@@ -0,0 +1,14 @@
+require "app/services/normalize_config"
+
+describe NormalizeConfig do
+  describe "#run" do
+    context "given a hash with keys containing capital letters" do
+      it "downcases the keys" do
+        config = { "Ruby" => { "Enabled" => true } }
+
+        expect(NormalizeConfig.run(config).keys).to eq ["ruby"]
```
Comment (thorncp) [inline spec/services/normalize_config_spec.rb]: What do you think about `expect(NormalizeConfig.run(config)).to have_key("ruby")`?

File: app/services/resolve_config_aliases.rb (hunk 11 above)
Comment (thorncp) [inline app/services/resolve_config_aliases.rb]: Another option could be `each_with_object`. A benefit would be being able to delete L23, since the method keeps its own accumulator.

File: app/models/hound_config.rb [hunk 28]
```diff
@@ -25,35 +28,29 @@ def fail_on_violations?
...
-    Config::Parser.yaml(file_content) || {}
-  end
-
-  def configured?(name)
-    key = normalize_key(name)
-    config = options_for(key)
-
-    enabled?(config)
+  def default_config
+    Linter::Collection.linter_names.each.with_object({}) do |name, config|
```
Comment (thorncp) [inline app/models/hound_config.rb]: Would it be beneficial for this generated config to be run through the normalizer and alias resolver? `Linter::Collection.linter_names` is including underscores, e.g. `coffee_script`, but config coming from the user is being resolved to `coffeescript`. Is the difference being resolved somewhere else?

Comment (jcmorrow) [inline app/models/hound_config.rb]: I don't think I understand the question. If the user provides a key `coffeescript` it's resolved as an alias for the true linter name: `coffee_script`, which is enabled by default. `coffee_script` is being passed in here and returns true as expected.

Comment (thorncp) [inline app/models/hound_config.rb]: Maybe I'm confused? Maybe a script would be a more clear representation:

```ruby
require "./config/environment"

class FakeCommit
  def file_content(*)
    <<~YAML
      ---
      coffee_script:
        enabled: false
    YAML
  end
end

commit = FakeCommit.new
hound_config = HoundConfig.new(commit)

puts hound_config.linter_enabled?("coffee_script") # => true
```

I think I'd expect it to return `false`. Is that correct?

Comment (jcmorrow) [inline app/models/hound_config.rb]: 😑

File: app/models/hound_config.rb [hunk 29]
```diff
@@ -12,11 +12,14 @@ class HoundConfig
   attr_reader_initialize :commit
 
   def content
-    @content ||= parse(commit.file_content(CONFIG_FILE))
+    @merged_config ||= default_config.deep_merge(resolved_aliases_config)
```
Comment (gylaz) [inline app/models/hound_config.rb]: We tend to name memoized variables after the method. Better yet, to add `_`, so `@_content`.

File: app/services/normalize_config.rb [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+    new(config).run
+  end
+
+  def initialize(config)
+    @config = config
+  end
+
+  def run
+    @config.each_with_object({}) do |(key, value), normalized_config|
+      normalized_key = normalize_key key
```
Comment (gylaz) [inline app/services/normalize_config.rb]: How about parens around `key`?

File: app/models/hound_config.rb (hunk 28 above)
Comment (thorncp) [inline app/models/hound_config.rb]: It seems to be because the order in `ResolveConfigAliases` is `"coffee_script" => "coffeescript"` (I haven't tried). Would it make sense to flip those?

Comment (jcmorrow) [inline app/models/hound_config.rb]: I think the aliases were just set up wrong:

`coffee_script` was aliasing to `coffeescript`, but it should be the other way around.

Comment (thorncp) [inline app/models/hound_config.rb]: :+1:

**Comment by jcmorrow** [comment]
@teoljungberg We're still feeling some pain from this and I'd like to merge it, but I don't want to do so if you feel uncomfortable with the approach. What do you think?

**Comment by teoljungberg** [comment]
@jcmorrow The thing I'm bothered by is the duplication between the various linter names and their aliases. Which are present in `ResolveConfigAliases` and `Linter::Base#linter_names`. But since they aren't connected at this point - I'm unsure it can be avoided.

I did make some improvements and refactorings to address the feedback and some of my own. But besides that - I'm fine merging this and moving on.

File: app/services/normalize_config.rb [hunk 31]
```diff
@@ -0,0 +1,23 @@
+class NormalizeConfig
+  def initialize(config)
+    @config = config
+  end
+
+  def run
+    @config.reduce({}) do |normalized_config, (key, value)|
```
Comment (gylaz) [inline app/services/normalize_config.rb]: @teoljungberg  Why did you change this from `each_with_object` to `reduce`? It seems like `each_with_object` is a better fit, since line 18 can be removed.

**Comment by jcmorrow** [comment]
@teoljungberg I totally understand your concern about duplicating the linter aliases. I've removed them from the linter classes as they are no longer necessary 👍

File: app/services/normalize_config.rb (hunk 31 above)
Comment (teoljungberg) [inline app/services/normalize_config.rb]: I find `reduce` easier to reason / read than `each_with_object` - if I'm alone in that camp I'm happy changing it back.
